File: cses/dynamic-programming/cc/coin_combinations1.cc
#include <iostream>
#include <vector>

int main() {
    int n, sum, mod = 1000000007;
    std::cin >> n >> sum;
    int coins[n];
    for (int i = 0; i < n; i++) {
        std::cin >> coins[i];
    }

    std::vector<int> ways(sum+1, 0);
    ways[0] = 1;

    for (int i = 1; i <= sum; i++) {
        for (int j = 0; j <n; j++) {
            if (i - coins[j] >= 0) {
                ways[i] = (ways[i]%mod + ways[i-coins[j]]%mod)%mod; // optimize this % using subtraction
            }
        }
    }

    std::cout << ways[sum] << '\n';
}

File: cses/dynamic-programming/java/coinCombinations1.java
import java.util.Scanner;

public class coinCombinations1 {
    public static void main(String[] a) {
        solve();
    }

    static void solve() {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int sum = sc.nextInt();
        int coins[] = new int[n];
        for (int i = 0; i < n; i++) {
            coins[i] = sc.nextInt();
        }
        sc.close();

        int ways[] = new int[sum+1];
        int mod = 1000000007;

        ways[0] = 1;
        for (int i = 1; i <= sum; i++) {
            for (int j = 0; j < n; j++) {
                int c = coins[j];
                if (i >= c) {
                    ways[i] += ways[i - c];
                    if (ways[i] >= mod) ways[i] -= mod;
                }
            }
        }

        System.out.println(ways[sum]);
    }
}
